src/lib.rs

```diff
@@ -494,3 +495,468 @@ mod muldiv_i16_tests {
+    //                *  vh   vl
+    //                ----------
+    // a0 =              vl * nl
+    // a1 =         vl * nh
+    // b0 =         vh * nl
+    // b1 =  + vh * nh
+    //       -------------------
+    //        c1h  c1l  c0h  c0l
+    //
+    // "a0" is optimized away, result is stored directly in c0.  "b1" is
+    // optimized away, result is stored directly in c1.
+    //
+
+    let mut c0 = v.lo() * n.lo();
+    let a1 = v.lo() * n.hi();
+    let b0 = v.hi() * n.lo();
+
+    // add the high word of a0 to the low words of a1 and b0 using c1 as
+    // scrach space to capture the carry.  the low word of the result becomes
+    // the final high word of c0
+    let mut c1 = c0.hi() + a1.lo() + b0.lo();
+
+    c0 = (c1.lo() << 64) | c0.lo();
+
+    // add the carry from the result above (found in the high word of c1) and
+    // the high words of a1 and b0 to b1, the result is c1.
+    c1 = v.hi() * n.hi() + c1.hi() + a1.hi() + b0.hi();
+
+    U256 { h: c1, l: c0 }
+}
+
+// based on Hacker's Delight p152
+
+fn u256_div_u128(num: U256, denom: u128) -> u128 {
+    assert!(denom > U64_MAX);
+
+    let mut c1 = num.h;
+    let mut c0 = num.l;
+
+    // count number of leading zeroes, we know they must be in the high
+    // part of denom since denom > G_MAXUINT64.
+    let s = (denom.hi() as u64).leading_zeros();
+
+    let v;
+    if s > 0 {
+        // normalize divisor and dividend
+        v = denom << s;
+        c1 = (c1 << s) | (c0.hi() >> (64 - s));
+        c0 <<= s;
+    } else {
+        v = denom;
+    }
+
+    let mut q1 = c1 / v.hi();
+    let mut rhat = c1.wrapping_sub(q1.wrapping_mul(v.hi()));
+
+    let mut cmp1 = (rhat.lo() << 64) | c0.hi();
+    let mut cmp2 = q1.wrapping_mul(v.lo());
+
+    while (q1 >> 64 != 0) || cmp2 > cmp1 {
+        q1 -= 1;
+        rhat += v.hi();
+        if rhat.hi() != 0 {
+            break;
+        }
+
+        cmp1 = (rhat.lo() << 64) | cmp1.lo();
+        cmp2 -= v.lo();
+    }
+
+    c1 = (c1.lo() << 64) | c0.hi();
+    c1 = c1.wrapping_sub(q1.wrapping_mul(v));
+
+    let mut q0 = c1 / v.hi();
+
+    rhat = c1.wrapping_sub(q0.wrapping_mul(v.hi()));
+
+    cmp1 = (rhat.lo() << 64) | c0.lo();
+    cmp2 = q0.wrapping_mul(v.lo());
+
+    while q0.hi() != 0 || cmp2 > cmp1 {
+        q0 -= 1;
+        rhat += v.hi();
+        if rhat.hi() != 0 {
+            break;
+        }
+
+        cmp1 = (rhat.lo() << 64) | cmp1.lo();
+        cmp2 -= v.lo();
+    }
+
+    q0 = ((q0.hi() + q1.lo()) << 64) | q0.lo();
+
+    q0
+}
+
+fn u128_scale_u128_unchecked(val: u128, num: u128, denom: u128, correct: u128) -> Option<u128> {
+    Some(u128_mul_u128(val, num))
+        .and_then(|r| u256_correct(r, correct))
+        .and_then(|r| {
+            if r.h >= denom as u128 {
+                None
+            } else {
+                Some(u256_div_u128(r, denom))
+            }
+        })
+}
+
+pub fn u128_scale(val: u128, num: u128, denom: u128, correct: u128) -> Option<u128> {
```

Comment:
Instead of having this as a public function, please implement the `MulDiv` trait. That's the public API of this crate :)

Reply:
This is still pending